### PR TITLE
Lock kinesis impressions

### PR DESCRIPTION
**Pull request by cavis**
Prevent firing duplicate downloads/impressions for up to 1 hour.  After that, the analytics-ingest-lambda needs to take some other action to prevent the reuse of a URL from inserting dups into BigQuery and re-pinging Adzerk/pingback urls.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,12 +86,24 @@ that as an IAB-2.0 complaint download.  For segments, we wait for _all_ the byte
...
 Since we might receive additional requests _after we've logged a download_, we
-can end up firing a whole bunch of downloads for the same listener-session-digest.  
-So it's important that they're considered idempotent in BigQuery, or some sort of
-unique constraint is applied later on.
+also lock the impression via a redis hash:
 
-TODO: introduce some sort of "this-was-already-logged" redis lock so we don't fire
-a bunch of duplicates into BigQuery.
+```
+redis:6379> HGETALL dtcounts:imp:<listener-session>/<digest>
```
Comment (cavis) [inline README.md]: Went with a hash because it seemed the most memory efficient?  Not entirely sure the 1-hour TTL is "right", but we'll find out.

Comment (svevang) [reply README.md]: Disregarding storage costs, would using the 24 hour rolling window as TTL remove all chance of dupe logging here ?

File: lib/redis.js [hunk 2]
```diff
@@ -99,6 +99,18 @@ module.exports = class Redis {
     return this.cmd('del', arguments)
   }
 
+  async lock(key, field, ttl) {
```
Comment (svevang) [inline lib/redis.js]: Neat use of redis to build a lock. Curious if DDB could function in the same way? e.g. using the `lsd` in ddb as the lock?

File: README.md (hunk 1 above)
Comment (cavis) [reply README.md]: Unfortunately no.  So the "listener-session" is an md5 of listenerId + url + dateInUTC.  But it's the date of the redirect, not the date of the download.

File: lib/redis.js (hunk 2 above)
Comment (cavis) [reply lib/redis.js]: Yeah, that's sort of what I'm thinking.  But we may just delete the cached-impression-urls in DDB to indicate we already processed them.

So this lock is a _bit_ redundant with that?  But my thought was that handling race conditions would be easier here with redis than DDB.  Then the ingest lambda doesn't have to deal with locking anything while it hits pingback urls or something.

**Pull request merged**